Add mTLS support.

r0mant opened the pull request:
This PR add support for client certificate authentication to Gravity web handlers and teaches "tele" to use Teleport-issued user certificates when talking to those APIs.

The high-level user-facing impact of this is that now "tele" can be used right after "tsh login" - "tele login" is no longer required.

Some details:

* Web handlers for pack/app/ops services can now authenticate incoming requests by x509 certificates provided by the clients. The way it works is:
  * All handlers are wrapped in Teleport's AuthMiddleware that looks for optional x509 cert in the HTTP request, extracts username from it and adds it to the request context.
  * Handlers then look if username is present in the request context and use it to get user info. Otherwise they fallback to regular auth methods such as bearer token or basic auth.

* All our handlers had exactly the same "needsAuth" logic so I factored it out into a dedicated "Authenticator" which all of them now use. No more code duplication!

* The "tele" client now knows how to search the standard Teleport's key store (~/.tsh) for the current cluster and user-certificates, and uses them when creating pack/app/ops clients which then triggers the authentication flow described above on the server side. The "old" token-based authentication still takes precedence though, for the time being for backwards compatibility.

* I felt like the "LocalEnvironment" was becoming too messy so I also factored out all credentials-related logic to a "CredentialsService" which is responsible to retrieving user credentials from different sources - Gravity config (which will become obsolete), Teleport key store or bolt key store.

There are still quite a few things outstanding so fully replacing "tele login" with "tsh login" is far from done, but this chunk should be fine to go in on its own.

References https://github.com/gravitational/gravity.e/issues/4198.

a-palchikov commented on lib/ops/resources/gravity/gravity_test.go:
```diff
@@ -51,9 +53,14 @@ var _ = check.Suite(&GravityResourcesSuite{
 	// create the ops client that uses admin agent creds
-	client, err := opsclient.NewBearerClient(s.server.URL, s.s.Creds.Password)
+	client, err := opsclient.NewBearerClient(s.server.URL, s.s.Creds.Password,
+		opsclient.HTTPClient(&http.Client{
```
I've this at least 3 times, `newInsecureHTTPClient`?

r0mant commented on lib/builder/builder.go:
```diff
@@ -124,6 +126,14 @@ func (c *Config) CheckAndSetDefaults() error {
 		c.GetRepository = getRepository
 	}
+	if c.Credentials == nil {
+		c.Credentials, err = localenv.NewCredentials(localenv.CredentialsServiceConfig{
```
This might be better, yes.

r0mant commented on lib/localenv/credentials.go:
```diff
@@ -0,0 +1,352 @@
+	}
+	if s.dbKeyStore != nil {
+		entries, err := s.dbKeyStore.GetLoginEntries()
+		if err == nil && len(entries) == 1 {
```
This is actually how it's worked before too, this logic was just ported from localenv as-is.

knisbet commented on lib/localenv/localenv.go:
```diff
@@ -438,9 +311,9 @@ func (env *LocalEnvironment) OperatorService(opsCenterURL string, options ...htt
-	return operator, trace.Wrap(err)
+	return env.OperatorService(defaults.GravityServiceURL,
+		httplib.WithLocalResolver(env.DNS.Addr()),
+		httplib.WithInsecure())
```
I realize not in scope for this PR, but I really need to work on getting rid of these hardcoded insecure calls everywhere.

knisbet commented on lib/ops/opshandler/opshandler.go:
```diff
@@ -2307,62 +2335,25 @@ func (s *WebHandler) needsAuth(fn ServiceHandle) httprouter.Handle {
 	if err != nil {
-		return nil, trace.Wrap(err, "credentials error")
+		log.WithError(err).Debug("Authentication error.")
+		return nil, trace.AccessDenied("bad username or password") // Hide the actual error.
```
When suppressing an error, I think logging of that error should not be debug, IE something high enough that it would actually be present in the logs. Possibly even with a correlation ID, where the correlation ID can be used to find the full error.

knisbet commented on lib/ops/opshandler/opshandler_test.go:
```diff
@@ -99,12 +101,21 @@ func (s *OpsHandlerSuite) SetUpTest(c *C) {
+		opsclient.HTTPClient(&http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true,
```
Nit: I usually try to generate an x509 cert inside the test suite, so that the tests don't run with x09 verification disabled. 

https://github.com/gravitational/satellite/blob/c383450edfb3dc4354be14d141dfd004b923e4c9/agent/server_test.go#L724-L778

r0mant replied:
Yeah, it's actually even easier as the test TLS server provides the underlying client which is already configured with proper certs/pool.

r0mant commented on lib/users/ap.go:
```diff
@@ -0,0 +1,47 @@
+		return "", trace.Wrap(err)
+	}
+	if trace.IsNotFound(err) {
+		return "", nil
```
That's how I had done it originally, but this method is called by the internal Teleport's middleware and it does not tolerate not handle not found errors atm. I'll add a comment here.

r0mant commented on lib/users/auth.go:
```diff
@@ -0,0 +1,190 @@
+//
+// The user is set in the request context by an authentication middleware that
+// extracts it from the verified client-supplied x509 certificate.
+func (a *authenticator) authenticateContext(r *http.Request) (storage.User, services.AccessChecker, error) {
```
Good call.

r0mant commented:
@a-palchikov @knisbet I've addressed your comments: https://github.com/gravitational/gravity/pull/638/commits/6ff41afec38ca8ceceb10e3d6054758c708c82a2.
